### PR TITLE
ninjabackend: Correct RPATH order

### PR DESCRIPTION
**Pull request by Finii**
[why]
If we build and test a library we need to make sure that we find the
currently build library object first, before an older system installed
one.
This can be broken if the library in question is installed in a custom
path, and another library we depend on also is installed there.

[how]
Just move the rpath to the current build artifacts to the front.

Solves #8030.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

### CODE REVIEW

**Comment by jpakkane** [comment]
This needs a test.

**Comment by Finii** [comment]
> This needs a test.

Is that a request in my direction? :->
The "setup" is relatively complicated, which is why I created a whole [how-to-reproduce repository](https://github.com/Finii/foo).
I can try of course, but I would rather leave this to ppl more experienced with the Meson test setup.

Please let me know how to proceed. Thank you, Fini

**Comment by mhier** [comment]
Any news on this? We really need this...

**Comment by jpakkane** [comment]
As said this needs a test. It should be implemented in `run_unittests.py´ under `LinuxlikeTests` and should do roughly this:

- have a custom .pc file for a dependency that just adds an rpath linker argument to a known value like ´/usr/lib/dummy`
- create a new project that imports said dependency, `PKG_CONFIG_PATH` needs to be set to the directory where your dummy .pc file is
- in the project build one shared lib and one executable that links to it
- it should also use the dummy dependency so its rpath flag is used
- build the project
- extract the rpath from the exe file, verify that both of the rpaths are there and that the build dir one (it always starts with `$ORIGIN`) comes before the system one

Overall the Meson file should look something like this:

```meson
project('rpathorder', 'c')

dummy_dep = dependency('rpathdummy')
l = shared_library('helper', 'helper.c')
executable('rpathtest', 'rpathtest.c', link_with: l, dependencies: dummy_dep)
```

**Comment by Finii** [comment]
* rebase on master
* added test

I added the test not in `linuxlike` but rather in `unit`, because `10 build_rpath` (which approx checks the same thing, just without pkgconfig) also resides there.

In fact I copied that test to make this new test. First I incorporated the pkgconfig test into `10 build_rpath` but that was not so nice when `@skipIfNoPkgconfig` needs to be supported.

Thank you, Fini

**Comment by Finii** [comment]
Force push: Correct whitespace error

File: run_unittests.py [hunk 1]
```diff
@@ -6733,6 +6733,31 @@ def test_build_rpath(self):
...
+        # C program RPATH
+        build_rpath = get_rpath(os.path.join(self.builddir, 'prog'))
+        self.assertEqual(build_rpath, '$ORIGIN/sub:/foo/bar:/foo/dummy')
+        self.install()
+        install_rpath = get_rpath(os.path.join(self.installdir, 'usr/bin/prog'))
+        self.assertEqual(install_rpath, '/baz:/foo/dummy')
+        # C++ program RPATH
+        build_rpath = get_rpath(os.path.join(self.builddir, 'progcxx'))
+        self.assertEqual(build_rpath, '$ORIGIN/sub:/foo/bar:/foo/dummy')
+        self.install()
```
Comment (jpakkane) [inline run_unittests.py:6757]: You are running install twice. Once is sufficient. So something like:

- build
- check C build dir rpath
- check C++ build dir rpath
- install
- check C install rpath
- check C++ install rpath

**Comment by Finii** [comment]
Removed superfluous second install.

Created PR #8222 where it is also removed from the testcase that has been used as copy source for this.

**Pull request merged**